### PR TITLE
Added autograd sampling op

### PR DESCRIPTION
**Pull request by abustamanteTT**
### Problem description
Sampling op did not previously work on autograd tensors

### What's changed
- Added wrapper to ttnn_fixed::sampling for autograd tensors

### Checklist
https://github.com/tenstorrent/tt-metal/actions/runs/17991867018
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tt-train/sources/ttml/ops/sampling_op.cpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+#include "autograd/graph.hpp"
+#include "autograd/graph_utils.hpp"
+
+namespace ttml::ops {
+
+autograd::TensorPtr ttml::ops::sample_op(
+    autograd::TensorPtr& t, float temperature, uint32_t seed, std::optional<autograd::TensorPtr&> logits_padding_mask) {
+    tt::tt_metal::Tensor sampled_tensor;
+
+    if (logits_padding_mask.has_value()) {
```
Comment (dmakoviichuk-tt) [inline tt-train/sources/ttml/ops/sampling_op.cpp]: could you use https://en.cppreference.com/w/cpp/utility/optional/transform.html
to avoid this if else check?

File: tt-train/sources/ttml/ops/sampling_op.cpp [hunk 2]
```diff
@@ -0,0 +1,35 @@
+#include "sampling_op.hpp"
+
+#include "autograd/auto_context.hpp"
+#include "autograd/graph.hpp"
+#include "autograd/graph_utils.hpp"
+
+namespace ttml::ops {
+
+autograd::TensorPtr ttml::ops::sample_op(
```
Comment (dmakoviichuk-tt) [inline tt-train/sources/ttml/ops/sampling_op.cpp]: dont need to add ttml::ops::, you are already inside of this namespace

File: tt-train/sources/ttml/ops/sampling_op.hpp [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+#include "autograd/tensor.hpp"
+#include "ttnn_fixed/trivial_ttnn_ops.hpp"
+
+namespace ttml::ops {
+
+autograd::TensorPtr sample_op(
+    autograd::TensorPtr& t,
+    float temperature,
+    uint32_t seed,
+    std::optional<autograd::TensorPtr&> logits_padding_mask = std::nullopt);
```
Comment (dmakoviichuk-tt) [inline tt-train/sources/ttml/ops/sampling_op.hpp]: I think optional doesn't support reference, btw I prefer to jsut pass const reference to the autograd::TensorPtr. It can be nullptr. YOu probably don't need optional.
and then you can get std::optional<tttnn::Tensor> value = logits_padding_mask ? logits_padding_mask->get_value() : nullopt_t;

Comment (abustamanteTT) [reply tt-train/sources/ttml/ops/sampling_op.hpp]: I think that the ternary operator doesn't like Tensor : nullopt; I tried that before the current if-else and I was getting an error about operand types being incompatible

File: tt-train/sources/ttml/ops/sampling_op.cpp (hunk 1 above)
Comment (abustamanteTT) [reply tt-train/sources/ttml/ops/sampling_op.cpp]: We're using C++20, no? std::optional::transform() is a C++23 feature

File: tt-train/sources/ttml/ops/sampling_op.hpp (hunk 3 above)
Comment (abustamanteTT) [reply tt-train/sources/ttml/ops/sampling_op.hpp]: Resolved, commit [0a7a3f8](https://github.com/tenstorrent/tt-metal/pull/29271/commits/0a7a3f8e815723bdaa635a95fbf0c6f65fdbfa01)

File: tt-train/sources/ttml/ops/sampling_op.cpp (hunk 2 above)
Comment (abustamanteTT) [reply tt-train/sources/ttml/ops/sampling_op.cpp]: Resolved, commit [0a7a3f8](https://github.com/tenstorrent/tt-metal/pull/29271/commits/0a7a3f8e815723bdaa635a95fbf0c6f65fdbfa01)

File: tt-train/sources/ttml/ops/sampling_op.cpp (hunk 1 above)
Comment (abustamanteTT) [reply tt-train/sources/ttml/ops/sampling_op.cpp]: Moot point, resolved in [0a7a3f8](https://github.com/tenstorrent/tt-metal/pull/29271/commits/0a7a3f8e815723bdaa635a95fbf0c6f65fdbfa01)

File: tt-train/sources/ttml/ops/sampling_op.hpp [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+#pragma once
+
+#include "autograd/tensor.hpp"
+#include "ttnn_fixed/trivial_ttnn_ops.hpp"
+
+namespace ttml::ops {
+
+autograd::TensorPtr sample_op(
+    autograd::TensorPtr& t, float temperature, uint32_t seed, const autograd::TensorPtr& logits_padding_mask);
```
Comment (dmakoviichuk-tt) [inline tt-train/sources/ttml/ops/sampling_op.hpp]: first input should be const reference

File: tt-train/sources/ttml/ops/sampling_op.hpp [hunk 5]
```diff
@@ -0,0 +1,15 @@
+// SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
```
Comment (dmakoviichuk-tt) [inline tt-train/sources/ttml/ops/sampling_op.hpp]: 2025

File: tt-train/sources/ttml/ops/sampling_op.cpp [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    auto out = autograd::create_tensor();
+    out->set_value(sampled_tensor);
+
+    autograd::GradFunction grad = [t, out]() {
+        // Argmax in sampling is non-differentiable; no gradient to propagate.
+        // TODO: replace argmax with softmax?
+        throw std::runtime_error("Sampling operation backward pass is not implemented.");
+    };
+
+    auto links = autograd::get_links(t);
```
Comment (dmakoviichuk-tt) [inline tt-train/sources/ttml/ops/sampling_op.cpp]: I'd pass both t and logits_padding_mask jsut for consistency.

File: tt-train/sources/ttml/ops/sampling_op.hpp (hunk 4 above)
Comment (dmakoviichuk-tt) [reply tt-train/sources/ttml/ops/sampling_op.hpp]: Also could you rename t to the logits because we have temperature which kind of 't' too :D

Comment (abustamanteTT) [reply tt-train/sources/ttml/ops/sampling_op.hpp]: Commit [3177374](https://github.com/tenstorrent/tt-metal/pull/29271/commits/31773741cf357b5a46cd4e685e4e40344c87e260)

File: tt-train/sources/ttml/ops/sampling_op.hpp (hunk 5 above)
Comment (abustamanteTT) [reply tt-train/sources/ttml/ops/sampling_op.hpp]: Commit [3177374](https://github.com/tenstorrent/tt-metal/pull/29271/commits/31773741cf357b5a46cd4e685e4e40344c87e260)

File: tt-train/sources/ttml/ops/sampling_op.cpp (hunk 6 above)
Comment (abustamanteTT) [reply tt-train/sources/ttml/ops/sampling_op.cpp]: Commit [e82b35c](https://github.com/tenstorrent/tt-metal/pull/29271/commits/e82b35cb5e228dfdbb6746d0d57556da826a0eb8)

File: tt-train/sources/ttml/ops/sampling_op.cpp [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    const autograd::TensorPtr& logits_padding_mask) {
+    auto sampled_tensor = ttnn_fixed::sample(
+        logits->get_value(),
+        temperature,
+        seed,
+        logits_padding_mask == nullptr ? std::nullopt
+                                       : std::optional<tt::tt_metal::Tensor>(logits_padding_mask->get_value()));
+
+    auto out = autograd::create_tensor();
+    out->set_value(sampled_tensor);
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/ops/sampling_op.cpp]: you can use

`auto out = autograd::create_tensor(sampled_tensor);`

**Review by rfurko-tt — APPROVED** [review]
LGTM, with a tiny nit

File: tt-train/sources/ttml/ops/sampling_op.cpp [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+    autograd::GradFunction grad = [logits, out]() {
```
Comment (rfurko-tt) [inline tt-train/sources/ttml/ops/sampling_op.cpp]: remove logits and out from context (you don't need them)

File: tt-train/sources/ttml/ops/sampling_op.cpp (hunk 7 above)
Comment (abustamanteTT) [reply tt-train/sources/ttml/ops/sampling_op.cpp]: Thank you! Commit [87135db](https://github.com/tenstorrent/tt-metal/pull/29271/commits/87135db0951f3d5c7c8c9e51208d0c3a287f2a44)

File: tt-train/sources/ttml/ops/sampling_op.cpp (hunk 8 above)
Comment (abustamanteTT) [reply tt-train/sources/ttml/ops/sampling_op.cpp]: Commit [87135db](https://github.com/tenstorrent/tt-metal/pull/29271/commits/87135db0951f3d5c7c8c9e51208d0c3a287f2a44)

**Review by dmakoviichuk-tt — APPROVED** [review]
